x/datadeal/keeper/deal.go

```diff
@@ -150,79 +136,82 @@ func (k Keeper) ListDeals(ctx sdk.Context) ([]types.Deal, error) {
+	if deal.GetStatus() != types.ACTIVE {
+		return sdk.Coin{}, sdkerrors.Wrapf(types.ErrInvalidStatus, "%s", deal.GetStatus())
 	}
 
-	dealAddress, err := types.AccDealAddressFromBech32(findDeal.GetDealAddress())
+	dealAddress, err := sdk.AccAddressFromBech32(deal.GetDealAddress())
 	if err != nil {
 		return sdk.Coin{}, err
 	}
 
-	err = k.isTrustedValidator(cert, findDeal)
-	if err != nil {
-		return sdk.Coin{}, err
+	if !k.isTrustedDataValidator(cert, deal) {
+		return sdk.Coin{}, types.ErrInvalidDataVal
 	}
 
-	totalAmount := findDeal.GetBudget().Amount.Uint64()
-	countOfData := findDeal.GetMaxNumData()
-	pricePerData := sdk.NewCoin(assets.MicroMedDenom, sdk.NewIntFromUint64(totalAmount/countOfData))
+	//TODO: Fields max_num_data and cur_num_data will be changed in next data datadeal model.
+	totalBudget := deal.GetBudget().Amount.ToDec()
+	maxNumData := sdk.NewIntFromUint64(deal.GetMaxNumData()).ToDec()
+	pricePerData := totalBudget.Quo(maxNumData).TruncateInt()
+
+	reward := sdk.NewCoin(assets.MicroMedDenom, pricePerData)
 
 	dealBalance := k.bankKeeper.GetBalance(ctx, dealAddress, assets.MicroMedDenom)
-	if dealBalance.IsLT(pricePerData) {
+	if dealBalance.IsLT(reward) {
 		return sdk.Coin{}, fmt.Errorf("deal's balance is smaller than reward")
 	}
 
-	coins := append(sdk.Coins{}, pricePerData)
+	coins := append(sdk.Coins{}, reward)
 
 	err = k.bankKeeper.SendCoins(ctx, dealAddress, seller, coins)
 	if err != nil {
 		return sdk.Coin{}, sdkerrors.Wrapf(types.ErrNotEnoughBalance, "The deal's balance is not enough to make deal")
 	}
 
-	k.SetDataCertificate(ctx, findDeal.GetDealId(), cert)
-	SetCurNumData(&findDeal)
+	k.SetDataCertificate(ctx, deal.GetDealId(), cert)
+	SetCurNumData(&deal)
 
-	if findDeal.GetCurNumData() == findDeal.GetMaxNumData() {
-		SetStatusCompleted(&findDeal)
+	if deal.GetCurNumData() == deal.GetMaxNumData() {
+		SetStatusCompleted(&deal)
 	}
 
-	k.SetDeal(ctx, findDeal)
-	return pricePerData, nil
+	k.SetDeal(ctx, deal)
+
+	return reward, nil
 }
 
-func (k Keeper) isDataCertDuplicate(ctx sdk.Context, cert types.DataValidationCertificate) error {
+func (k Keeper) isDuplicatedData(ctx sdk.Context, cert types.DataValidationCertificate) bool {
 	store := ctx.KVStore(k.storeKey)
 	dataCertKey := types.GetKeyPrefixDataCertificate(cert.UnsignedCert.GetDealId(), cert.UnsignedCert.GetDataHash())
 
-	if store.Has(dataCertKey) {
-		return sdkerrors.Wrapf(types.ErrDataAlreadyExist, "data %s is already exist.", dataCertKey)
-	}
-
-	return nil
+	return store.Has(dataCertKey)
 }
 
-func (k Keeper) isTrustedValidator(cert types.DataValidationCertificate, findDeal types.Deal) error {
+func (k Keeper) isTrustedDataValidator(cert types.DataValidationCertificate, deal types.Deal) bool {
 	validator := cert.UnsignedCert.GetDataValidatorAddress()
-	trustedValidators := findDeal.GetTrustedDataValidators()
+	trustedValidators := deal.GetTrustedDataValidators()
+
+	if len(trustedValidators) == 0 || trustedValidators == nil {
```

Comment:
I think if trustedValidators are nil, len(trustedValidators) will return 0.

```suggestion
	if len(trustedValidators) == 0 {
```